src/ethereum/cancun/vm/instructions/system.py

```diff
@@ -515,17 +515,24 @@ def selfdestruct(evm: Evm) -> None:
     beneficiary_balance = get_account(evm.env.state, beneficiary).balance
     originator_balance = get_account(evm.env.state, originator).balance
 
-    # First Transfer to beneficiary
-    set_account_balance(
-        evm.env.state, beneficiary, beneficiary_balance + originator_balance
-    )
-    # Next, Zero the balance of the address being deleted (must come after
-    # sending to beneficiary in case the contract named itself as the
-    # beneficiary).
-    set_account_balance(evm.env.state, originator, U256(0))
+    if (
+        originator in evm.env.state._created_accounts
+        or originator != beneficiary
+    ):
+        # First Transfer to beneficiary
+        set_account_balance(
+            evm.env.state,
+            beneficiary,
+            beneficiary_balance + originator_balance,
+        )
+        # Next, Zero the balance of the address being deleted (must come after
+        # sending to beneficiary in case the contract named itself as the
+        # beneficiary).
+        set_account_balance(evm.env.state, originator, U256(0))
```

Comment:
Something about these `if` conditions is bothering me, but I can't quite put my finger on it.

I haven't thought to deeply about the logic, but would something like this be easier to understand?

```python
# Increase beneficiary's balance
move_ether(
    evm.env.state,
    originator,
    beneficiary,
    originator_balance,
)

if originator in evm.env.state._created_accounts:
    # Zero the balance of the address being deleted (must come after
    # sending to beneficiary in case the contract named itself as the
    # beneficiary).
    set_account_balance(evm.env.state, originator, U256(0))

    # Register account for deletion
    evm.accounts_to_delete.add(originator)
````

Reply:
You're right! This is in fact simpler to understand. Updated